source/beamline/BioXAS/BioXASBiStateGroup.cpp

```diff
@@ -0,0 +1,363 @@
+		// those that can move can also be stopped.
+
+		QList<AMControl*> children = childControls();
+
+		if (children.count() > 0) {
+			bool childrenValid = true;
+			bool childrenStoppable = true;
+
+			for (int i = 0, count = children.count(); i < count && childrenValid && childrenStoppable; i++) { // we want to stop if we come across any invalid children or children that 'should' be stoppable but aren't.
+				AMControl *child = children.at(i);
+
+				if (child) {
+					if (child->canMove() && !child->canStop()) // we want to know if we have any children that can move but can't be stopped. Potential problem!
+						childrenStoppable = false;
+
+				} else {
+					childrenValid = false; // we want to know if we have any children that aren't valid. Another potential problem.
+				}
+			}
+
+			result = childrenValid && childrenStoppable;
+		}
+	}
+
+	return result;
+}
+
+void BioXASBiStateGroup::updateConnected()
+{
+	bool connected = false;
+
+	// Iterate through list of children, finding out if
+	// all are connected.
+
+	QList<AMControl*> children = childControls();
+
+	if (children.count() > 0) {
+
+		bool childrenConnected = true;
+
+		for (int i = 0, count = children.count(); i < count && childrenConnected; i++) { // we want to stop if any one child isn't connected.
+			AMControl *child = children.at(i);
+
+			if (child && child->isConnected())
+				childrenConnected = true;
+			else
+				childrenConnected = false;
+		}
+
+		connected = childrenConnected;
+	}
+
+	setConnected(connected);
+}
+
+void BioXASBiStateGroup::updateMoving()
+{
+	// Iterate through list of children, finding out if
+	// any are moving.
+
+	QList<AMControl*> children = childControls();
+
+	if (children.count() > 0) {
+		bool moving = false;
+
+		for (int i = 0, count = children.count(); i < count && !moving; i++) { // we want to stop if any one child IS moving.
+			AMControl *child = children.at(i);
+
+			if (child && child->isMoving())
+				moving = true;
+			else
+				moving = false;
+		}
+
+		setIsMoving(moving);
+	}
+}
+
+bool BioXASBiStateGroup::addBiStateControl(AMControl *control, double state1Value, double state2Value)
+{
+	bool result = false;
+
+	if (control) {
+		controlState1ValueMap_.insert(control, state1Value);
+		controlState2ValueMap_.insert(control, state2Value);
+
+		addChildControl(control);
+
+		result = true;
+	}
+
+	return result;
+}
+
+bool BioXASBiStateGroup::removeBiStateControl(AMControl *control)
+{
+	bool result = false;
+
+	if ( control && (controlState1ValueMap_.contains(control) || controlState2ValueMap_.contains(control)) ) {
+		controlState1ValueMap_.remove(control);
+		controlState2ValueMap_.remove(control);
+
+		removeChildControl(control);
+
+		result = true;
+	}
+
+	return result;
+}
+
+bool BioXASBiStateGroup::clearBiStateControls()
+{
+	bool result = false;
+
+	QList<AMControl*> children = childControls();
+
+	if (children.count() > 0) {
+		foreach (AMControl *child, childControls())
```

Comment:
Since you made `removeBiStateControl()` a boolean function, you could use that boolean here inside this loop.